Ci GitHub actions

woky opened the pull request:
This adds GitHub Actions as CI in addition to Drone. There's artifact upload missing, but it can still be merged to run on PRs/pushes. Compared to Drone, this CI runs everything for PRs as well.

Commit:
ci: Add GitHub Actions CI

Everything is documented in comments. This commit does not change main
CI (bors.toml). GitHub Actions CI workflow in this commit can run
alongside Drone without any interference.

Later commit will switch main CI to Drone. If something goes wrong,
revert the later commit instead of this one.

Commit:
itests: Get rid of mount_dir

Misleading and almost unused.

Commit:
itests: Make test launch script more portable

- Do not install Pyenv and custom Python in Drone launcher. This change
  depends on latest buildenv[1] that has Pyenv already installed.
  Also put Python version selection into .python-version file which is
  picked up by Pyenv by default. This file is intentionally missing
  because tests run fine with Python 3.6 in Ubuntu bionic. In case you
  require specific Python version, just put it into .python-version.

- Do not launch Python from hardcoded path. Use "pyenv exec" instead.
  This makes it possible to use Pyenv in custom location (using
  PYENV_ROOT and PATH variables), as it is done in buildenv[1].

[1] https://github.com/rchain/buildenv

zsluedem commented on integration-tests/run_tests:
```diff
@@ -0,0 +1,9 @@
+docker build -t $DEFAULT_IMAGE .
+
+export TMPDIR=${TMPDIR-/tmp}
+python3 -m pytest -v --mount-dir=$TMPDIR "$@"
```
Since we remove the `get_absolute_path_for_mounting` , is `--mount-dir`necessary?

woky replied:
Done.

zsluedem commented on .github/print-integration-test-selection:
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/python3
```
file name `print-integration-test-selection` -> `print-integration-test-selection.py`?
or it has to be `print-integration-test-selection`

woky replied:
Some use extensions for executable scripts, some don't and rely on a) executable bit and b) shebang line. For instance, `pip` and `pipenv` subscribe to the latter religion:
```
% ls -l /usr/bin/pip3
-rwxr-xr-x 1 root root 366 Nov  3 13:08 /usr/bin/pip3
% head -1 /usr/bin/pip3
#!/usr/bin/python
```

zsluedem commented on .github/workflows/continuous-integration.yml:
```diff
@@ -0,0 +1,244 @@
+          - test_slash_GHOST_disobeyed
+          - test_storage
+          - test_synchrony_constraint
+          - test_wallets
```
~~So if I add a new test , I should put the test name here, right? 
Is there a more intelligent way  to do that?~~

zsluedem replied:
After reading your comments in discord https://discordapp.com/channels/391435113415573504/417429638642532363/657998758474416139, I totally get it~
If I don't specify the test here ,all other tests would run in `REMAINER` and it would not run in parallel. Am I right?

woky replied:
You're right.

zsluedem reviewed (commented):
Very fantastic and great job! Leave some comments.

tgrospic commented on .github/shared.py:
```diff
@@ -0,0 +1,27 @@
+import yaml
+
+WORKFLOW_PATH = '.github/workflows/continuous-integration.yml'
+REMAINDER = 'REMAINDER'
```
It would be great if you can add some description about `REMAINDER` and why/how it's used.

woky replied:
I've added comment elsewhere. Please review again from the workflow file.

tgrospic commented on .github/init-environment:
```diff
@@ -0,0 +1,76 @@
+#!/bin/bash
+set -eu -o pipefail
+
+CACHE_ROOT=/var/cache/rchain-build
```
Short description of how cache is working would be very helpful.

woky replied:
Done.

zsluedem reviewed (approved):
LGTM

woky commented:
@tgrospic @nzpr @zsluedem I've made more changes. Now it's properly documented and it fully replaces Drone. Please review. Tested extensively in past week. See: https://github.com/woky/rchain/actions.

Commit:
Make all LMDB map sizes 10TB by default

The size denotes memory map size limit. When the limit is reached, LMDB
operations will fail. We do not have any estimates on how much memory is
going to be needed for running RChain network, so just put the limit
high enough.

For information about what a "map size" in LMDB is, read documentation
on map_size argument at https://lmdb.readthedocs.io/en/release/#environment-class

nzpr commented on bors.toml:
```diff
@@ -1,14 +1,14 @@
 
 # Number of seconds from when a merge commit is created to when its statuses must pass.
-timeout_sec = 14400 #4h
+timeout_sec = 5400 #4h
```
#1.5h ?

woky replied:
Fixed.

nzpr reviewed (approved):
Great job, I love comments!

woky commented:
bors try

woky commented:
bors r+

woky commented:
bors r-

woky commented:
bors r+

woky commented:
Well, we're hitting the exact same problem as in https://forum.bors.tech/t/resource-not-accessible-by-integration/408/9, see https://app.bors.tech/repositories/5710/log.

woky commented:
bors try

woky commented:
bors try-

woky commented:
bors r+

woky commented:
OK. Let's try to merge this by reducing it to just introducing GHActions as a side CI and preserving Drone as it is, i.e. removing commit 9d4dd321 (ci: Switch to GitHub Actions as main CI), so that GHActions gets enabled, and merge CI switch in a followup PR. Sigh...

woky commented:
bors r+

woky commented:
https://github.community/t5/GitHub-Actions/GitHub-action-Resource-not-accessible-by-integration/m-p/19950/highlight/true#M75
>OAuth tokens and GitHub Apps are restricted from editing the main.workflow file. This means it isn't possible to merge a Pull Request via an Action if it updates the main.workflow file I'm afraid.

IOW, we can't use Bors to update `.github/workflow/continuous-integration.yml`. But note, we most probably can use Bors to modify `.github/a-script-called-from-workflow-doing-nefarious things`. `¯\_(ツ)_/¯`

Commit:
WIP: ci: test

woky commented:
bors r+

woky commented:
bors r-
